picard/ui/options/plugins.py

```diff
@@ -59,46 +90,131 @@ def __init__(self, parent=None):
             self.loader = "file://%s"
         self.ui.install_plugin.clicked.connect(self.open_plugins)
         self.ui.folder_open.clicked.connect(self.open_plugin_dir)
-        self.ui.plugin_download.clicked.connect(self.open_plugin_site)
+        self.ui.reload_available_plugins.clicked.connect(self.reload_available_plugins)
         self.tagger.pluginmanager.plugin_installed.connect(self.plugin_installed)
+        self.tagger.pluginmanager.plugin_updated.connect(self.plugin_updated)
+        self.ui.plugins.header().setStretchLastSection(False)
+        self.ui.plugins.header().setResizeMode(0, QtGui.QHeaderView.Stretch)
+        self.ui.plugins.header().setResizeMode(1, QtGui.QHeaderView.Stretch)
+        self.ui.plugins.header().resizeSection(2, 100)
+        self.ui.plugins.setSortingEnabled(True)
 
     def load(self):
+        self.ui.details.setText("<b>" + _("No plugins installed.") + "</b>")
         plugins = sorted(self.tagger.pluginmanager.plugins, cmp=cmp_plugins)
         enabled_plugins = config.setting["enabled_plugins"]
-        firstitem = None
+        available_plugins = dict([(p.module_name, p.version) for p in
+                                  self.tagger.pluginmanager.available_plugins])
+        installed = []
         for plugin in plugins:
-            enabled = plugin.module_name in enabled_plugins
-            item = self.add_plugin_item(plugin, enabled=enabled)
-            if not firstitem:
-                firstitem = item
-        self.ui.plugins.setCurrentItem(firstitem)
+            if plugin.module_name in enabled_plugins:
+                plugin.enabled = True
+            if plugin.module_name in available_plugins.keys():
+                latest = available_plugins[plugin.module_name]
+                if latest.split('.') > plugin.version.split('.'):
+                    plugin.new_version = latest
+                    plugin.can_be_updated = True
+            item = self.add_plugin_item(plugin)
+            installed.append(plugin.module_name)
+
+        for plugin in sorted(self.tagger.pluginmanager.available_plugins, cmp=cmp_plugins):
+            if plugin.module_name not in installed:
+                plugin.can_be_downloaded = True
+                item = self.add_plugin_item(plugin)
+
+        self.ui.plugins.sortByColumn(0, QtCore.Qt.AscendingOrder)
+        self.ui.plugins.setCurrentItem(self.ui.plugins.topLevelItem(0))
+
+    def reload_available_plugins(self):
```

Comment:
This should block clicking the button while the download is in progress, otherwise it's possible to click the button fast enough to cause multiple requests at the same time, which then lead to duplicate entries in the plugin list.